res/controllers/Traktor-Kontrol-S4-MK3.js

```diff
@@ -0,0 +1,1949 @@
+                this.output(engine.getValue(this.group, this.outKey));
+            });
+        }
+    }
+}
+
+class SamplerButton extends Button {
+    constructor(options) {
+        super(options);
+        if (this.number === undefined || !Number.isInteger(this.number) || this.number < 1 || this.number > 64) {
+            throw Error("SamplerButton must have a number property of an integer between 1 and 64");
+        }
+        this.group = "[Sampler" + this.number + "]";
+        this.outConnect();
+    }
+    input(pressed) {
+        if (!this.shifted) {
+            if (pressed) {
+                if (engine.getValue(this.group, "track_loaded") === 0) {
+                    engine.setValue(this.group, "LoadSelectedTrack", 1);
+                } else {
+                    engine.setValue(this.group, "cue_gotoandplay", 1);
+                }
+            }
+        } else {
+            if (pressed) {
+                if (engine.getValue(this.group, "play") === 1) {
+                    engine.setValue(this.group, "play", 0);
+                } else {
+                    engine.setValue(this.group, "eject", 1);
+                }
+            } else {
+                if (engine.getValue(this.group, "play") === 0) {
+                    engine.setValue(this.group, "eject", 0);
+                }
+            }
+        }
+    }
+    // This function is connected to multiple Controls, so don't use the value passed in as a parameter.
+    output() {
+        if (engine.getValue(this.group, "track_loaded")) {
+            if (engine.getValue(this.group, "play")) {
+                this.send(this.color + this.brightnessOn);
+            } else {
+                this.send(this.color + this.brightnessOff);
+            }
+        } else {
+            this.send(0);
+        }
+    }
+    outConnect() {
+        if (undefined !== this.group) {
+            this.outConnections[0] = engine.makeConnection(this.group, "play", this.output.bind(this));
+            this.outConnections[1] = engine.makeConnection(this.group, "track_loaded", this.output.bind(this));
+        }
+    }
+}
+
+class IntroOutroButton extends PushButton {
+    constructor(options) {
+        super(options);
+        if (this.cueBaseName === undefined || typeof this.cueBaseName !== "string") {
+            throw Error("must specify cueBaseName as intro_start, intro_end, outro_start, or outro_end");
+        }
+        this.outKey = this.cueBaseName + "_enabled";
+        this.outConnect();
+    }
+    unshift() {
+        this.inKey = this.cueBaseName + "_activate";
+    }
+    shift() {
+        this.inKey = this.cueBaseName + "_clear";
+    }
+    output(value) {
+        if (value) {
+            this.send(this.color + this.brightnessOn);
+        } else {
+            this.send(0);
+        }
+    }
+}
+
+class Pot extends Component {
+    constructor(options) {
+        super(options);
+        this.firstValueReceived = false;
```

Comment:
please document why this is required again.

Reply:
This is required so the software state is initialized to the hardware state on startup.

Reply:
I forgot how softtakeover works exactly so I can't comment on that further.

Reply:
For now... I'll read up on that tomorrow.

Reply:
I understand soft-takeover now. The solution is not elegant but it works I guess.